### ISSUE TITLE
Per-archive configuration / stored filenames renaming

### ISSUE THREAD

**Issue by lazuratus** [hasse69/rar2fs#60]
_This is a 2 part feature request..._

## Part 1: Per-archive configuration
**It would be very useful to be able to have per-archive configuration, so that some archives could be treated differently. For example:**

If the default mount uses `--seek-length=1` and we have those files:
```
tv.show.s01.extras.rar
tv.show.s01.extras.r01
tv.show.s01.extras.r02
tv.show.s01.extras.rar2fs
```
the `.rar2fs` file could contain different configuration:
```
--seek-length=0
```
This would allow rar2fs to look at all volumes for this set of archives only, and would use `--seek-length=1` for everything else.

The `.rar2fs` file should not be visible in the mounted filesystem, just like the archives are right now.

P.S. I used archive.name for the `.rar2fs` file, but rar2fs could maybe look for any `.rar2fs` file, like `_configuration.rar2fs` or `_renaming.rar2fs`... just an idea...

## Part 2: Stored filenames renaming
**When using rar2fs with PLEX/Kodi or other software, sometimes the archives contains files that are not recognized as the proper season/episode/movie name. By having some way to change the stored name using that same `.rar2fs` file, it would allow the media player to see a different name, and properly recognize the episode.**

If for example, the archive contains:
```
tv.show.s01.extra-01.mkv
tv.show.s01.extra-02.mkv
```
The `.rar2fs` file could tell how to show those files:
```
tv.show.s01.extra-01.mkv|tv.show.s00e02.mkv
tv.show.s01.extra-02.mkv|tv.show.s00e05.mkv
```
or, using some sort of prefix?
```
--rename=tv.show.s01.extra-01.mkv|tv.show.s00e02.mkv
--rename=tv.show.s01.extra-02.mkv|tv.show.s00e05.mkv
```
This would get those extras recognized as special episodes from thetvdb.com when scanned with PLEX or other software.

**Comment by hasse69** [comment]
Doable. I will see what can be done (and when).

**Comment by nubje** [comment]
+1 for part 2 - sounds really usable.

**Comment by hasse69** [comment]
Regarding part 2, does it necessarily have to be a file per archive, would a file like .rarrename in the mount point root be able to describe all the names to be replaced by an alternative one?
The reason why I ask is because it would be a lot more simple (and effcient) to implement population of a simple hash table and then key into this table to find a file alternative name instead of having to open a file for each archive access or search the entire mount point for config files.

 I was thinking of a format like, e.g:

```
-tv.show.s01.extra-01.mkv
+tv.show.s00e02.mkv
-tv.show.s01.extra-02.mkv
+tv.show.s00e05.mkv
```

**Comment by lazuratus** [comment]
One file per archive would be a lot more practical (considering, I might have over 500 files to be renamed, scattered over many sub-directories), but maybe that could come later if/when part 1 can be implemented?

For now, this would work for me. I would keep one file per archive in the format you mentioned in separate sub-directories, and would just write a script to generate a `.rarrename` in the mount point root by concatenating all the `.rarrename` files in those sub-directories.

Would be really nice tho, if empty lines and comments were allowed:

    # TV.Show.S01.Extras-GRP
    -tv.show.s01.extra-01.mkv
    +tv.show.s00e02.mkv
    -tv.show.s01.extra-02.mkv
    +tv.show.s00e05.mkv

    # 2nd.TV.Show.S02.Extras-GRP
    -2nd.tv.show.s02.extra-01.mkv
    +2nd.tv.show.s00e03.mkv

**Comment by hasse69** [comment]
Ok, how you merge or create the file in question is entirely up to you. I am looking for the most common use case and having to add a file to each archive seems like even more work to me?
But I have been giving this some more thought and to prepare for possible future options per archive I propose instead a .rarsettings in the root with a format like:

```
:<archive name without extension>
-tv.show.s01.extra-01.mkv
+tv.show.s00e02.mkv
-tv.show.s01.extra-02.mkv
+tv.show.s00e05.mkv
```

Of course empty lines and comments, being # or whatever should be supported.

**Comment by lazuratus** [comment]
For me, it seems a LOT easier to manage with separate file for each archive. If for example, I add a new show with an EXTRAS folder, I can just generate a new `.rarsettings` inside that EXTRAS folder (using a context menu option on the archive itself), add the renames I want and optionally the options I want, and I know that folder is taken care of. Also, if I delete any EXTRAS folder, the `.rarsettings` file for that folder is removed as well.

If for example, I have 100 EXTRAS folders, each containing on average 5 media files, that's approximately 500 files that potentially need the renaming feature so that PLEX can correctly match them. Using only one file, means having over 1000 lines of rename instructions in one file, making it really hard to keep track of everything.

All this is irrelevant though. I think you should go with whatever is easier/faster for you to implement. Like I said before, I (or anyone else who wants to do the same) can just generate the `.rarsettings` file in the root of the mount point from `.rarsettings` files in any of the EXTRAS sub-directories.

P.S. Maybe let users choose filename if not using the default name? (example: `--config-file=.rar2fs-settings`)

**Comment by hasse69** [comment]
Ok, I will see what I can do. But you also need to consider the case that there are multiple archives in the same folder, then the config file still needs to cover more than one archive. The problem with having more than one file is the collection process. However, having one file also had its drawbacks since if an archive is added or the config file is changed it would have to be reloaded. Very well...

**Comment by lazuratus** [comment]
If it would be possible to have one config per archive, it could still be possible to use `:<archive name without extension>` if more than one archive is present in the same folder... or omitted if only one archive is present...

Another possibility would be using `archive name.rar2fs-settings` (or whatever extension is established). This way, there would be 2 setting files if there are 2 archives in the same folder.

**Comment by hasse69** [comment]
Would it be ok if the entry was using a complete file name, eg. [xxx.yyy.zzz.part001.rar], or [foobar.rar] because it is a real mess to extract the basename without first having to read the header and analyse the archive due to the extremely "open" new format (i.e. new style volumes does not even need the 'part' section and can be called basically anything). Basically it would be the first volume file or the one single file

**Comment by nubje** [comment]
I am not opposed to have a single settings file the root of the mount point, in that way it would be easier to keep track of it.

Would it make sense to specify the full path to the file, in case there are situations where the archives could have the same name, but different content?
```
:path/to/file.rar
[rename]
...
```

**Comment by lazuratus** [comment]
Yes, using the complete file name is fine by me! The only reason I mentioned `:<archive name without extension>` was because that's what you wrote in your example :) for me, it makes no difference.

**Comment by lazuratus** [comment]
I was thinking about it a little more, and if altering the path would be possible using the one config file in the root of the mount, I would definitely prefer this method. By alter path, I mean:

	:tv_shows/show/EXTRAS/<archive name>
	[rename]
	-tv_shows/show/EXTRAS/tv.show.s01.extra-01.mkv
	+tv_shows/show/Season 0/tv.show.s00e02.mkv
	-tv_shows/show/EXTRAS/tv.show.s01.extra-02.mkv
	+tv_shows/show/Season 0/tv.show.s00e05.mkv

Or using relative paths for the renames:

	:tv_shows/show/EXTRAS/<archive name>
	[rename]
	-tv.show.s01.extra-01.mkv
	+../Season 0/tv.show.s00e02.mkv
	-tv.show.s01.extra-02.mkv
	+..Season 0/tv.show.s00e05.mkv

The reason this would be useful, is that PLEX ignores EXTAS folders, but would scan Season 0 as Specials, making if very convenient to organize archived shows, that scan flawlessly in PLEX.

---

If this would be somehow possible using the one config per archive (using relative paths), even better.

	:<archive name>
	[rename]
	-tv.show.s01.extra-01.mkv
	+../Season 0/tv.show.s00e02.mkv
	-tv.show.s01.extra-02.mkv
	+..Season 0/tv.show.s00e05.mkv

**Comment by hasse69** [comment]
I have been giving this a few thoughts too. The idea of having a configuration that can specify per archive behavior is good. That is for sure something that is needed. However, the renaming I am more reluctant to accept as a rar2fs feature :( The rar2fs file system was made to mimic the exact result of an archives being extracted in the current file system layout. Renaming and faking of filenames sounds like the task for another file system that can be stacked on top of rar2fs. Also symbolic links could be used to solve the problem you describe I think?

**Comment by lazuratus** [comment]
I understand perfectly your point of view. But I've tried looking for other file systems that I could stack on top and found none... hence the request.

Currently, the best solution was to manually create symbolic links in a different location, and use Mergerfs to merge the symlinks with the rar2fs mount. This is a very long and annoying process (especially when I move archives, I need to re-create the symlinks)

One thing I didn't think of doing before, is creating something similar to those config files (the rename part), and write a script to automate the process of creating mass symbolic links using those config files.

I guess the thing that I really need the most from this request is the per-archive behavior, for those few archives containing multiple files (to allow `--seek-length=0` only when needed).

**Comment by hasse69** [comment]
Here is a patch that can be used to test the functionality. Some more work is needed but it should be good enough for a trial. Note that the patch is rebased to latest master and is depending on a bug-fix made just minutes ago. The config file should be placed in the root of the source folder and be named `.rarconfig`.
Only one option is currently supported, and the syntax is:

```
[<archive name> | </path_to_archive/archive_name]
        seek-length=<N>
...
```

As you can see, only a simple/base name or a complete path can be given depending on if it there are collisions and multiple copies with the same name. Also note that for volumes the name must be that of the first file in the set. Anything not matching the pattern above is simply ignored, so garbage or comments of any shape or form should simply pass unnoticed. Please feedback anything you see is not working as expected.

[patch_v1.txt](https://github.com/hasse69/rar2fs/files/917957/patch_v1.txt)

**Comment by lazuratus** [comment]
Tried it with 2 different archives and so far so good! Gonna try to somehow automate the process with a script, so I can test pretty much all that can be tested. Gonna get back with results in a week or so...

**Comment by hasse69** [comment]
Thanks! If you are interested I have a patch for addding support for password too. The seek length and password are most likely the primary options that needs to be separated per archive.

**Comment by hasse69** [comment]
I briefly looked at a rename kind of feature while at it. Still do not like it but if it would make life easier for some I will add it.  Quick question, I have two syntax proposals atm:

```
alias = /name^/new-name
```
or 
```
alias-from = /name
alias-to = /new-name
```

The drawback with first alternative is that it needs a delimiter, in this example I used `^` since it is very unlikely a filename will include such. The `/` character cannot be used since RAR files can include directories and the complete path inside the archive (starting with `/`) needs to be specified.
The drawback with the second alternative is that it needs to be parsed in pairs, thus two consecutive `alias-from` entries will not work (the second will hide the first) and vice versa.

Which one would you prefer? The first one is a lot easier to implement ;)
Worse case scenario with the first alternative is that the delimiter is used in a filename, but in that case we can add an `alias-delim` option to be able to change delimiter per archive too. But to be honest I do not think that will ever be needed.

**Comment by lazuratus** [comment]
First one is perfect!

I'm guessing this would go right under the `[<archive_name>]` section?

    [<archive name>]
        seek-length = 0
        alias = tv.show.s01.extra-01.mkv^tv.show.s00e02.mkv
        alias = tv.show.s01.extra-02.mkv^tv.show.s00e05.mkv

Also, if possible, it might be nice to be able to have whitespace on any side of the delimiter...

    alias-delim = |
    alias = tv.show.deleted-scenes.mkv    | tv.show.s00e02.mkv
    alias = tv.show.behind-the-scenes.mkv | tv.show.s00e05.mkv

**Comment by hasse69** [comment]
Maybe | is a better default delimiter then since you used that in your example? I rather avoid having to implement alias-delim, that it is more of an emergency escape route if needed later. White spaces can be anywhere except in the filename right now.

**Comment by hasse69** [comment]
Here is a new patch based on 240ea3d84be1e2503a9c4dec2a3552539ca0c741
Checkout the `rarconfig.example` file for the syntax of `alias`. Note that `alias` entries must always start with a `/` since it is the expressed as an absolute path relative to the to the root of the archive.

[patch_v3.txt](https://github.com/hasse69/rar2fs/files/941660/patch_v3.txt)

**Comment by nubje** [comment]
I've tried the `alias` on ~40 archives, and so far it's working flawless - great work!

Is the `.rarconfig` only read on mount, or is there any way refresh it after updating it?

**Comment by lazuratus** [comment]
Working, but not 100% for me... took me a while to figure out why it's misbehaving sometimes (`.rarconfig` being completely ignored), but I finally found it. I'm mounting rar2fs using Autofs, and if I try to access the rar2fs mount (`/mnt/rar2fs/tv_shows`) after a reboot, the config file is ignored.

But, if I first go into the samba share mount (`/mnt/smb/tv_shows`), and then go into `/mnt/rar2fs/tv_shows`, the `.rarconfig` seems to work flawlessly.

Not sure if it's rar2fs or Autofs related and not too sure how to figure this out or how to fix it.

**Comment by hasse69** [comment]
Thanks for testing!

@nubje Next patch will implement SIGHUP which means `kill -HUP <rar2fs-pid>` will re-read the configuration (and invalidate the cache). Tell me if that is not good enough.

@lazuratus My bad. I most likely try to read the configuration to soon before fuse has actually mounted/accessed the source folder which means autofs might not yet have had a chance to expose it yet. Will try to fix it in next patch.

**Comment by hasse69** [comment]
And here it is...
Note that the `alias` delimiter has now changed from `|` to `,` 

[patch_v4.txt](https://github.com/hasse69/rar2fs/files/948793/patch_v4.txt)

**Comment by lazuratus** [comment]
Thank you! Problem somehow fixed, but weird behavior... I can now access the `/rar2fs/` mount directly, but, all the archives for which I've used `alias`, now show the renamed file and the original one as well... not sure if this is intended. Also, the file with the original filename is unplayable.

**Comment by hasse69** [comment]
Does not sound like fixed to me then...
Can you try to not use autofs and see if double display of files are still there? Really weird because that should not be possible! I do not see that problem so it must be ralated to autofs somehow?

**Comment by hasse69** [comment]
@lazuratus Can you try patch_v3 again with your autofs workaround to see if you see the same problems?
Maybe it is related to patch_v4 somehow?

**Comment by lazuratus** [comment]
No duplicate files, no seeking issues, and no video glitches with v3.

**Comment by hasse69** [comment]
Try this one, it is patch 3 with some parts of patch 4 that should not affect playback at all.
(actually I am rather surprised v4 did too but I need to check more carefully)

[patch_v5.txt](https://github.com/hasse69/rar2fs/files/948889/patch_v5.txt)

**Comment by hasse69** [comment]
Use this patch instead

[patch_v6.txt](https://github.com/hasse69/rar2fs/files/949028/patch_v6.txt)

**Comment by lazuratus** [comment]
Tested v6 briefly, and so far no issues. The rar2fs mount can be accessed directly, the --seek-length option works, alias works, no more duplicate files, seeking works, and no apparent video glitches.

**Comment by hasse69** [comment]
Nothing works!? That sounds like a bummer...
Please report back more details a soon as you have a chance to test!

**Comment by lazuratus** [comment]
Nevermind, I feel stupid... while playing with my install script, I accidentally added a typo... I had:

    whet https://github.com/hasse69/rar2fs/files/949028/patch_v6.txt

instead of:

    wget https://github.com/hasse69/rar2fs/files/949028/patch_v6.txt

Guess that's what you get when you rush things before going to work...  :expressionless:

Everything is working again!

**Comment by hasse69** [comment]
Ok, great! I will hold this commit until I push the patch for the new UnRAR source version(s) to master.
Then I will make a patch release of 1.24.0 and put this on on top of that...

**Comment by hasse69** [comment]
Patch has been merged to master. Please test and report back.

**Comment by lazuratus** [comment]
Seems to be working just fine! Will do more tests before the end of Thursday, if anything I'll report back.

**Comment by hasse69** [comment]
Any update here?

**Comment by lazuratus** [comment]
Just finished testing, and everything is working perfectly so far. I'm currently creating `.rar2fs-config` files in the folder where the archive is, and I wrote a batch file to merge it into `.rarconfig` in the root of the mount.

Tested with `seek-length` and `alias` only, since I don't have any password protected archives, but the 35 or so archives (~5 with `seek length`, and ~30 with `alias`) work as expected. I'll start using this version from now on, and I'll continue adding more archives. If I encounter any issues, I'll report them.

Thanks again for those 2 features!
